### PR TITLE
Annotate Worker and Workflow TwillRunnables to allow submitting them as k8s jobs

### PR DESCRIPTION
**Pull request by CuriousVini**
image

### CODE REVIEW

File: cdap-master-spi/src/main/java/io/cdap/cdap/master/spi/twill/ContainerRuntime.java [hunk 1]
```diff
@@ -16,10 +16,17 @@
...
 /**
- * Constants used for twill.
+ * Annotation to denote {@link TwillRunnable} can run on all supported container runtime environments.
  */
-public class TwillConstants {
-  // Configuration used to determine whether to use V1Job to launch twill runnables.
-  public static final String PROGRAM_RUNTIME_ENV = "program.runtime.env";
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ ElementType.TYPE })
+public @interface ContainerRuntime {
```
Comment (CuriousVini) [inline cdap-master-spi/src/main/java/io/cdap/cdap/master/spi/twill/ContainerRuntime.java]: Calling it `ContainerRuntime` to represent runnables that can run on any container runtime supported by CDAP. 
Also open to any other suggestions on the name.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillPreparer.java [hunk 2]
```diff
@@ -527,9 +528,15 @@ public TwillController start(long timeout, TimeUnit timeoutUnit) {
       StatefulRunnable statefulRunnable = statefulRunnables.get(mainRuntimeSpec.getName());
       Type resourceType = statefulRunnable == null ? V1Deployment.class : V1StatefulSet.class;
 
-      if (runnableConfigs.getOrDefault(mainRuntimeSpec.getName(),
-                                       Collections.emptyMap()).containsKey(TwillConstants.PROGRAM_RUNTIME_ENV)) {
-        resourceType = V1Job.class;
+      // Check if ContainerRuntime annotation is present on the twill runnable. If its present, that twill runnable
+      // would be submitted as kubernetes job
+      ClassLoader contextClassLoader = Thread.currentThread().getContextClassLoader();
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillPreparer.java]: The call maybe null. So better be

```
ClassLoader cl = Optional.ofNullable(Thread.currentThread().getContextClassLoader())
  .orElse(getClass().getClassLoader());
```

**Review by chtyim — COMMENTED** [review]
Everything runs in container. I think it is better be something that indicates how the task terminate. Whether it should be restarted or not

**Comment by CuriousVini** [comment]
> Everything runs in container. I think it is better be something that indicates how the task terminate. Whether it should be restarted or not

Yea that makes more sense. Using `CompletableTwillRunnable`

File: cdap-master-spi/src/main/java/io/cdap/cdap/master/spi/twill/CompletableTwillRunnable.java [hunk 3]
```diff
@@ -16,10 +16,17 @@
...
 /**
- * Constants used for twill.
+ * Annotation to denote a {@link TwillRunnable} that completes.
  */
-public class TwillConstants {
-  // Configuration used to determine whether to use V1Job to launch twill runnables.
-  public static final String PROGRAM_RUNTIME_ENV = "program.runtime.env";
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ ElementType.TYPE })
+public @interface CompletableTwillRunnable {
```
Comment (chtyim) [inline cdap-master-spi/src/main/java/io/cdap/cdap/master/spi/twill/CompletableTwillRunnable.java]: I would just call this `Completable`.

File: cdap-master-spi/src/main/java/io/cdap/cdap/master/spi/twill/Completable.java [hunk 4]
```diff
@@ -16,10 +16,17 @@
...
 /**
- * Constants used for twill.
+ * Annotation to denote a {@link TwillRunnable} that completes.
  */
-public class TwillConstants {
-  // Configuration used to determine whether to use V1Job to launch twill runnables.
-  public static final String PROGRAM_RUNTIME_ENV = "program.runtime.env";
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ ElementType.TYPE })
+public @interface Completable {
```
Comment (chtyim) [inline cdap-master-spi/src/main/java/io/cdap/cdap/master/spi/twill/Completable.java:31]: Shall we add this to the `SparkTwillRunnable` as well?

**Review by chtyim — APPROVED** [review]

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillPreparer.java [hunk 5]
```diff
@@ -527,9 +528,16 @@ public TwillController start(long timeout, TimeUnit timeoutUnit) {
...
 
-      if (runnableConfigs.getOrDefault(mainRuntimeSpec.getName(),
-                                       Collections.emptyMap()).containsKey(TwillConstants.PROGRAM_RUNTIME_ENV)) {
-        resourceType = V1Job.class;
+      // Check if ContainerRuntime annotation is present on the twill runnable. If its present, that twill runnable
+      // would be submitted as kubernetes job
+      ClassLoader contextClassLoader = Optional.ofNullable(Thread.currentThread().getContextClassLoader())
+        .orElse(getClass().getClassLoader());
+      for (Annotation annotation : contextClassLoader
+        .loadClass(getMainRuntimeSpecification(twillSpec.getRunnables())
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillPreparer.java]: The `mainRuntimeSpec` is already stored in a local variable.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillPreparer.java [hunk 6]
```diff
@@ -527,9 +528,16 @@ public TwillController start(long timeout, TimeUnit timeoutUnit) {
...
+                     .getRunnableSpecification().getClassName()).getAnnotations()) {
+        if (annotation.annotationType().getName().equals(Completable.class.getCanonicalName())) {
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillPreparer.java]: Shouldn't compare with the name. Use `instanceof Completable` to check.

File: cdap-master-spi/src/main/java/io/cdap/cdap/master/spi/twill/Completable.java (hunk 4 above)
Comment (CuriousVini) [reply cdap-master-spi/src/main/java/io/cdap/cdap/master/spi/twill/Completable.java:31]: For now, we only support worker and workflow. Spark still requires additional changes. I was thinking to tag it in separate PRs.

**Review by chtyim — APPROVED** [review]

**Pull request merged**